applications/OptimizationApplication/python_scripts/utilities/response_function_implementor.py

```diff
@@ -0,0 +1,194 @@
+from abc import ABC
+from abc import abstractmethod
+
+import KratosMultiphysics as Kratos
+from KratosMultiphysics.OptimizationApplication.optimization_info import OptimizationInfo
+from KratosMultiphysics.OptimizationApplication.responses.response_function import ResponseFunction
+from KratosMultiphysics.OptimizationApplication.utilities.helper_utilities import ContainerVariableDataHolderUnion
+
+class ResponseFunctionImplementor(ABC):
+    def __init__(self, response_name: str, optimization_info: OptimizationInfo):
+        self.__response_function: ResponseFunction = optimization_info.GetOptimizationProcess(ResponseFunction, response_name)
+        self.__name = response_name
+
+        self.__initial_response_value = None
+        self._optimization_info = optimization_info
+        self._key_prefix = f"problem_data/response_data/{self.__response_function.GetModelPart().FullName()}/{self.GetName()}"
+
+    def GetName(self) -> str:
+        return self.__name
+
+    def GetValue(self, solution_step_index: int = 0) -> float:
+        key = f"{self._key_prefix}/value"
+        if solution_step_index == 0 and not self._optimization_info.HasValue(key):
+            self._optimization_info.SetValue(key, self.__response_function.CalculateValue())
+        return self._optimization_info.GetValue(key, solution_step_index)
+
+    def GetSensitivity(self, sensitivity_variable, sensitivity_data_container: ContainerVariableDataHolderUnion, solution_step_index: int = 0):
```

Comment:
I find it confusion that the method is called "Get.." (just as the one above, "GetValue") but it is not returning anything. What the method does is coping data to the optimization info. I think that should be reflected in the naming. 
For example: def CopySensitivitiesToOptimizationInfo

Reply:
I agree the Get Function should return the object. I don't know if the logic for solution_step_index == 0 should be inside this function. With less hidden logic, it would be easier to use and extend in the future. 

Maybe you can use GetSensitivity to simply Get and through an error if the data doesn't exist? What do you think? Of course, it leads to changes to move add CalculateSensitivity to some external place.

Reply:
Lets change the name to `CalculateSensitivity`

Reply:
done :)